modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11yChecker.ts

```diff
@@ -0,0 +1,398 @@
+ *
+ * The implementation is based on Queue. Each new scheduled callback is added
+ * end of the queue, the scheduler works with the main callback which is
+ * responsible for emptying the queue and executing the queue callbacks in the
+ * work loop respecting the frame deadline.
+ *
+ * Main thread
+ * ------------------------------------------------------------------------->
+ *
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ * |    Busy    | || || || |  Idle  |            |  Busy  |    | Idle |
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ *
+ *                         +----------+ +---+                  +-----------+
+ *                         | Job Exec.| |   |   Yield to host  | Job Exec. |
+ *                         +----------+ +---+ <--------------> +-----------+
+ *                        |                  |
+ *                        +------------------+
+ *                              Deadline
+ *
+ * When the main thread is idle the main callback is called to empty the queue
+ * then a work loop is executed to ensure that we can execute more than one
+ * callback until the defined deadline but it is interrupted to avoid loss
+ * of frame and at the end of each execution it defines the next frame phase
+ * that avoids abusing cycles and yields up control of the main thread to
+ * the host. The main callback will be rescheduled again until the queue is
+ * empty.
+ */
+class Scheduler<T> {
+	private taskIdCounter: number = 0;
+	private readonly queue: Queue<Task<T>>;
+	private currentTask: Task<T> | undefined = undefined;
+	private scheduledHostHandle: number = 0;
+	private isCallbackScheduled: boolean = false;
+	private isPerformingWork: boolean = false;
+
+	/**
+	 * The frame phases can be recognized as lanes which are a space of time
+	 * where the scheduler yields up the main thread to the host in that time
+	 * interval before taking up again.
+	 */
+	private nextFramePhase: number = 0;
+
+	/**
+	 * The yield interval is defined based on the RAIL model, this value is
+	 * static and means that we yield 100ms to the host and define the deadline
+	 * to maximize the chances of hitting 60 fps.
+	 */
+	private readonly yieldInterval: number = 100;
+	private deadline: number = 0;
+
+	constructor() {
+		this.queue = new Queue();
+	}
+
+	cancelHostCallback() {
+		window.cancelIdleCallback(this.scheduledHostHandle);
+	}
+
+	hasCallback(selector: Selector<Task<T>>) {
+		return this.queue.has(selector);
+	}
+
+	private async flushWork() {
+
+		// We'll need a host callback the next time work is scheduled.
+
+		this.isCallbackScheduled = false;
+
+		this.isPerformingWork = true;
+		try {
+			return await this.workLoop();
+		}
+		finally {
+			this.currentTask = undefined;
+			this.isPerformingWork = false;
+		}
+	}
+
+	private requestHostCallback() {
+
+		// We should consider something alternative because
+		// requestIdleCallback is not supported in Safari.
```

Comment:
In its first iteration, this is developer-mode-only, so we can simply feature-test and just won't have this running in Safari at all.

As a second step, integrated inside a Content Editor (CKEditor), the checks will be manually triggered, so there should be no need for a scheduler.

Reply:
Oh I imagined that inside a Content Editor we would also have the idea of real-time, although maybe it doesn’t even make sense even the granular analysis 🤔. Ok, so I'm going to keep that.

Reply:
User-facing features and usage would be up to the teams. If ContentEditor wants to track content for a11y, it would be up to them to figure out the proper workflow. 

Our main concern here is to help make sure the UIs we ship out of the box (the Liferay core apps) are accessible. So, we're targeting the code and apps our colleagues work, not what our customers will write :)

Reply:
Makes sense!